nixos/lib/test-driver/test-driver.py

```diff
@@ -1,1017 +1,70 @@
-def subtest(name: str) -> Iterator[None]:
-    with log.nested(name):
-        try:
-            yield
-            return True
-        except Exception as e:
-            log.log(f'Test "{name}" failed with error: "{e}"')
-            raise e
+import os
 
-    return False
+import ptpython.repl
 
+from logger import Logger
+from driver import Driver
 
 if __name__ == "__main__":
     arg_parser = argparse.ArgumentParser()
+    arg_parser.add_argument(
+        "-S",
+        "--start-scripts",
+        default=os.environ.get("startScripts"),
+        required=True,
+        nargs="+",
+        help="start scripts for participating virtual machines",
+    )
+    arg_parser.add_argument(
+        "-V",
+        "--vlans",
+        required=True,
+        default=os.environ.get("vlans"),
+        nargs="+",
+        type=int,
+        help="vlans to span by the driver",
+    )
     arg_parser.add_argument(
         "-K",
         "--keep-vm-state",
         help="re-use a VM state coming from a previous run",
         action="store_true",
     )
-    (cli_args, vm_scripts) = arg_parser.parse_known_args()
-
-    log = Logger()
+    arg_parser.add_argument(
+        "-I",
+        "--interactive",
+        help="drop into a python repl and run the tests interactively",
+        action="store_true",
+    )
+    arg_parser.add_argument(
+        "test-script",
+        default=os.environ.get("testScript"),
+        help="the test script to run",
+        type=argparse.FileType('r'),
+    )
 
-    vlan_nrs = list(dict.fromkeys(os.environ.get("VLANS", "").split()))
-    vde_sockets = [create_vlan(v) for v in vlan_nrs]
-    for nr, vde_socket, _, _ in vde_sockets:
-        os.environ["QEMU_VDE_SOCKET_{}".format(nr)] = vde_socket
+    args = arg_parser.parse_args()
 
-    machines = [
-        create_machine({"startCommand": s, "keepVmState": cli_args.keep_vm_state})
-        for s in vm_scripts
-    ]
-    machine_eval = [
-        "{0} = machines[{1}]".format(m.name, idx) for idx, m in enumerate(machines)
-    ]
-    exec("\n".join(machine_eval))
+    logger = Logger()
+    if not args.keep_vm_state:
+        logger("Machine state will be reset. To keep it, pass --keep-vm-state")
 
-    @atexit.register
-    def clean_up() -> None:
-        with log.nested("cleaning up"):
-            for machine in machines:
-                if machine.pid is None:
-                    continue
-                log.log("killing {} (pid {})".format(machine.name, machine.pid))
-                machine.process.kill()
-            for _, _, process, _ in vde_sockets:
-                process.terminate()
-        log.close()
+    driver = Driver(
+        logger,
+        args.start_scripts,
+        args.test_script,
+        args.keep_vm_state
+    )
 
-    tic = time.time()
-    run_tests()
-    toc = time.time()
-    print("test script finished in {:.2f}s".format(toc - tic))
+    if not args.interactive():
+        tic = time.time()
+        driver.start_all()
```

Comment:
Do we actually want this in non interactive mode? I see that basically all scripts include `start_all()` at the top and I can see there could be use cases where the test script needs finer control over how machines (anc eventually `vlans`) are started up.

I adopted this from previous code, which might have been code smell / dead code anyways.

Please confirm my analysis.